lib/interface_quda.cpp

```diff
@@ -5244,6 +5732,119 @@ int computeGaugeFixingFFTQuda(void *gauge, const unsigned int gauge_dir, const u
   return 0;
 }
 
+void contractFTQuda(void **prop_array_flavor_1, void **prop_array_flavor_2, void **result,
+		    const QudaContractType cType, void *cs_param_ptr, const int src_colors,
+		    const int *X, const int *const source_position,
+		    const int n_mom, const int *const mom_modes, const QudaFFTSymmType *const fft_type)
+{
+  profileContractFT.TPSTART(QUDA_PROFILE_TOTAL);
+  profileContractFT.TPSTART(QUDA_PROFILE_INIT);
+
+
+  // create ColorSpinorFields from void** and parameter
+  auto cs_param = (ColorSpinorParam *)cs_param_ptr;
+  const size_t nSpin = cs_param->nSpin;
+  const size_t src_nColor = src_colors;
+  cs_param->location = QUDA_CPU_FIELD_LOCATION;
+  cs_param->create = QUDA_REFERENCE_FIELD_CREATE;
+  
+  // The number of complex contraction results expected in the output
+  size_t num_out_results = nSpin * nSpin;
+
+  //FIXME can we merge the two propagators if they are the same to save mem?
+  // wrap CPU host side pointers
+  std::vector<ColorSpinorField*> h_prop1, h_prop2;
+  h_prop1.reserve(nSpin*src_nColor);
+  h_prop2.reserve(nSpin*src_nColor);
+  for(size_t i=0; i<nSpin*src_nColor; i++) {
+    cs_param->v = prop_array_flavor_1[i];
+    h_prop1.push_back(ColorSpinorField::Create(*cs_param));
+    cs_param->v = prop_array_flavor_2[i];
+    h_prop2.push_back(ColorSpinorField::Create(*cs_param));
+  }
+  
+  // Create device spinor fields
+  ColorSpinorParam cudaParam(*cs_param);
+  cudaParam.create = QUDA_NULL_FIELD_CREATE;
+  cudaParam.location = QUDA_CUDA_FIELD_LOCATION;
+  cudaParam.gammaBasis = QUDA_DEGRAND_ROSSI_GAMMA_BASIS; // not relevant for staggered
+  cudaParam.setPrecision(cs_param->Precision(), cs_param->Precision(), true);
+  
+  std::vector<ColorSpinorField *> d_prop1, d_prop2;
```

Comment:
Move to stack

Reply:
fixed in 4fa8f65983905f955fa1dabad997b25e447653df